### DISCUSSION TITLE
A question about PolycrystalColoringIC

### DISCUSSION THREAD

**Discussion by Clark-domain** [idaholab/moose#19125] [Q&A Modules: Phase field] [answered]
Hi, MOOSE team,

I wanted to use the block parameter for the PolycrystalColoringIC in my work and found it was not working.

There was a warning sign saying block was not used. 

How can I set the block id for PolycrystalColoringIC? Actually, when I used the PolycrystalVoronoiVoidIC, it worked, and there was no warning.

Is it possible the code did not enable setting block id PolycrystalColoringIC?

Thanks very much,

Kaijie

[ICs]
  [./PolycrystalICs]
    [./PolycrystalColoringIC] #PolycrystalColoringIC not working?
      polycrystal_ic_uo = voronoi
      block = '1 2 3'   
    [../] 
  [../]
[]
![image](https://user-images.githubusercontent.com/74393696/137569749-805b7e99-cf51-4106-b87b-3926c51c5580.png)

**Comment by SudiptaBiswas** [comment]
The  `PolycrystalColoringIC` assigns grains based on the information provided by the `polycrystal_ic_uo`, that's why it's not using the block information you provided. You will have to specify the blocks in the UserObject for this.

**Reply by Clark-domain** [reply]
I just tried the following UserObjects, but still, there were warnings about the unused block. It was not working. 


[UserObjects]
  [./voronoi]
    type = PolycrystalVoronoi
    grain_num = 10
    rand_seed = 12444
    block = '1 2 3' 
  [../]
  [./grain_tracker]
    type = GrainTracker
    threshold = 0.2
    connecting_threshold = 0.08
    compute_halo_maps = true # Only necessary for displaying HALOS
    block = '1 2 3' 
  [../]
[]

![image](https://user-images.githubusercontent.com/74393696/137962606-5fe7e0f5-5388-4434-aa14-38db682e80ea.png)

**Comment by GiudGiud** [comment]
I had a look and it seems the parameter is not transmitted from the action to the ICs.
@SudiptaBiswas could you please have a look at this branch I made and see if that should work?
I can make a PR from there. 
https://github.com/GiudGiud/moose/blob/PR_action_crystal/modules/phase_field/src/action/PolycrystalVoronoiVoidICAction.C

@KaijieNing do you need boundary restriction too or just block restriction?

**Reply by Clark-domain** [reply]
Great! I likely need boundary restriction later. I appreciate it if you could solve both of them.

**Reply by dschwen** [reply]
@GiudGiud you edited the wrong class. See below.

**Reply by GiudGiud** [reply]
how did I manage that... I edited like 6 of these IC actions. 
Ok i ll fix that up

**Reply by GiudGiud** [reply]
oh I edited that one as well already, i just posted the wrong link

**Reply by GiudGiud** [reply]
Added boundary restriction, made the PR https://github.com/idaholab/moose/pull/19194

**Comment by dschwen** [comment]
We'll need to add the `block` parameter to the [PolycrystalColoringICAction](https://github.com/idaholab/moose/blob/next/modules/phase_field/src/action/PolycrystalColoringICAction.C). Should be pretty straight forward!

**Reply by Clark-domain** [reply]
Thanks so much. PolycrystalColoringICAction was not working instead of PolycrystalVoronoiVoidICAction, as far I know.

**Comment by GiudGiud** [comment, marked as answer]
Ok with the merging of https://github.com/idaholab/moose/pull/19194 this should be fixed. Thanks for reporting this!

**Reply by Clark-domain** [reply]
**Thanks very much.
I updated MOOSE and implemented the following commands. But after compiling my application, it was not working. Is there anything wrong when I did the update?**

```
(moose) [moose@localhost moose]$ **git rebase origin/master**
error: cannot rebase: You have unstaged changes.
error: Please commit or stash them.
(moose) [moose@localhost moose]$ **git rebase origin/master**
error: cannot rebase: You have unstaged changes.
error: Please commit or stash them.
(moose) [moose@localhost moose]$ **git stash**
Saved working directory and index state WIP on master: 59fb6b81d3 Merge commit '579cc6274bfa1525b7962796270dfcdbb1bb27b0'
(base) [moose@localhost moose]$ **git pull --rebase --autostash**
Successfully rebased and updated refs/heads/master.
(base) [moose@localhost moose]$ **git stash pop**
On branch master
Your branch is up to date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   modules/phase_field/test/tests/rigidbodymotion/polycrystal_action.i
	modified:   modules/tensor_mechanics/test/tests/global_strain/global_strain_disp.i
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	coupled_pressure_test1.i
	examples/a_MyCourse_Final/
	modules/combined/examples/publications/rapid_dev/fig7b-1.i
	modules/combined/examples/publications/rapid_dev/fig8-1.i
no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (a16e1507545f4303d6c58d3ed114911cfd84983c)
(base) [moose@localhost moose]$ **git status**
On branch master
Your branch is up to date with 'origin/master'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	coupled_pressure_test1.i
	examples/a_MyCourse_Final/
	modules/combined/examples/publications/rapid_dev/fig7b-1.i
	modules/combined/examples/publications/rapid_dev/fig8-1.i

nothing added to commit but untracked files present (use "git add" to track)
```

**Reply by GiudGiud** [reply]
did you re-compile? 
What you did with git updated the source code, but it did not update the executable

**Reply by Clark-domain** [reply]
Yes, I recompiled it and passed all the tests. It showed the latest MOOSE version. But I checked the folder and found PolycrystalColoringIC was not updated.
See below the three snapshots. 

![image](https://user-images.githubusercontent.com/74393696/139494537-6e8806da-52ea-4e73-8c30-e8b100203b90.png)

![image](https://user-images.githubusercontent.com/74393696/139494059-bac48d91-fed8-4458-8527-9a7ad29f6a23.png)

![image](https://user-images.githubusercontent.com/74393696/139494327-844b4b01-1601-4abe-93b5-17bdb4f8433c.png)

**Reply by GiudGiud** [reply]
oh it s because you pulled in the master branch. The changes I merged in arent in master yet. They are still in devel or next.

**Reply by GiudGiud** [reply]
oh looks like the changes are in master now
https://github.com/idaholab/moose/commits/master

**Reply by GiudGiud** [reply]
but the one you pulled was literally two commits before my PR got in master. If you repeat the maneuver now it should work

**Reply by Clark-domain** [reply]
Yes. You were right. I repeated it and it works now. Thanks very much.
